File: sock-merchant/sock-merchant.go
package sockmerchant

func sockMerchant(ignored int32, colors []int32) int32 {
	freqByColor := make(map[int32]int32)
	for _, color := range colors {
		if freqByColor[color] == 0 {
			freqByColor[color] = 1
		} else {
			freqByColor[color]++
		}
	}

	var pairCount int32 = 0
	for _, freq := range freqByColor {
		var pairableCount int32
		if freq%2 == 0 {
			pairableCount = freq
		} else {
			pairableCount = freq - 1
		}

		pairCount += (pairableCount / 2)
	}

	return pairCount
}

File: sock-merchant/SockMerchant.java
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

public class SockMerchant {

  public static int sockMerchant(int ignored, List<Integer> colors) {
    Map<Integer, Long> freqByColor = colors
      .stream()
      .collect(
            Collectors.groupingBy(
                Function.identity(),
                HashMap::new,
                Collectors.counting()
            )
        );

    int pairCount = 0;
    for (long freq : freqByColor.values()) {
      long pairableCount = freq %2 == 0
        ? freq
        : freq - 1;

      pairCount += (pairableCount / 2);
    }

    return pairCount;
  }

}
